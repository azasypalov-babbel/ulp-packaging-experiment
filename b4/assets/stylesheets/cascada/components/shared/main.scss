@use "sass:math";
// Layout

%responsive-container {
  margin: 0 auto;
  width: 90%;

  @include media(medium) {
    max-width: 940px;
  }
  @include media(large) {
    // fixed width for large and beyond
    max-width: none;
    width: 940px;
  }
  @include media(extra-large) {
    width: 1170px;
  }
}

// Typography

%trainer-page-text {
  font-size: math.div(16, 15) * 1em;
  margin-bottom: rem(24px);
  text-align: center;
}

%centered-columns {
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  flex-direction: column;
}

// UI

%bordered-box {
  border-radius: $border-radius;
  background-color: $white;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.16),
              0 0 2px 0 rgba(0, 0, 0, 0.12),
              inset 1px 1px 0 0 rgba(0, 0, 0, 0.08);
  color: $space-gray;
}

// Cascada override for Safari in iOS 10

.btn--primary,
.btn {
  display: inline-block;
}
