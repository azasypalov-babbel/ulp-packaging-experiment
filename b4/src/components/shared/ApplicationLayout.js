import styled, { css } from 'styled-components';
import { isiOS, isWebview } from '../../lib/features';

const fixNavigationBar = css`
  & .loy .loy-cascada-navbar {
    position: fixed;
    top: 0;
    padding-top: env(safe-area-inset-top);

    &:after {
      top: initial;
    }
  }
  & > main {
    margin-top: calc(${(props) => props.theme.size.navbar.xsmall.height} + env(safe-area-inset-top));
  }
`;

const disableSelection = css`
  user-select: none;
  -webkit-touch-callout: none;
`;

const ApplicationLayout = styled.div`
  display: flex;
  flex-direction: column;

  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    height: 100vh; /* IE11 does not respect min-height for calculating flex */
  }

  min-height: 100vh;
  ${isiOS() && css`
    min-height: -webkit-fill-available; /* To reduce the page size if possible when software keyboard is open */
  `}

  ${isWebview() && fixNavigationBar};
  ${isWebview() && disableSelection};

  & > main {
    position: relative;
    flex: 1 0 auto; /* main should be least the height of it's content */

    display: flex;

    & > * {
      flex: 1 1 100%; /* make children 100% width, and shrink if screen width is too small */
    }
  }

  .hide-for-tiny {
    @media (max-width: 413px) {
      display: none;
    }
  }

  .fixed-for-small {
    @media (max-width: 767px) {
      position: fixed;
    }
  }

  .hide-for-small {
    @media (max-width: 767px) {
      display: none;
    }
  }
`;

export default ApplicationLayout;
