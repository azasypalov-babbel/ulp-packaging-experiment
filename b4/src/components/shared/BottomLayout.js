import React, { useEffect, useLayoutEffect, useState } from 'react';
import { getDisplayName } from '../getDisplayName';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createPortal } from 'react-dom';
import useResizeObserver from './hooks/useResizeObserver';
import { getBottomInset, scrollIntoSafeArea, setBottomInset } from './scrollIntoSafeArea';
import { connect } from 'react-redux';

/**
 * This area is for any elements that are fixed to the bottom of the page
 * Give components `order` css property to organise them.
 */

const mountPointId = 'bottom-layout-mount-point';

export const StyledBottomLayout = styled.div`
  position: relative;

  @supports (position: sticky) {
    position: sticky;
    height: ${(props) => props.internalHeight}px;
  }

  bottom: 0;
  left: 0;
  right: 0;

  flex: 0 0 auto;
  z-index: 99;

  pointer-events: none;

  & > * {
    @supports (position: sticky) {
      position: absolute;
    }

    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    flex-direction: column;

    pointer-events: all;
  }
`;

export const bottomLayoutConfig = {
  enableScrollAdjustment: true,
  enableShrinking: false
};

const BottomLayoutComponent = ({ children, id }) => {
  const [internalHeight, setInternalHeight] = useState(0);

  const onResize = ({ height }) => {
    const isGrowing = height > internalHeight;

    if (isGrowing) {
      /**
       * InternalHeight is used to reserve space in the DOM.
       * It can only increase in height, to avoid making the page jump.
       */
      setInternalHeight(height);
    } else if (height !== 0 && bottomLayoutConfig.enableShrinking) {
      setInternalHeight(height);
    }

    const shouldAdjustScroll = height > getBottomInset();

    /**
     * Height of this element is used to calculate a safe area for scrolling
     * ie. so that content is not covered by anything in BottomLayout
     */
    setBottomInset(height);

    if (shouldAdjustScroll && bottomLayoutConfig.enableScrollAdjustment) {
      /**
       * Delay scroll until after next paint so that elements are in their final place
       */
      requestAnimationFrame(() => scrollIntoSafeArea());
    }
  };

  useEffect(() => {
    /**
     * Reset internal height when a the trainer changes or sequence starts or ends
     */
    setInternalHeight(0);
  }, [id]);

  const elementRef = useResizeObserver(onResize);

  return (
    <StyledBottomLayout internalHeight={internalHeight}>
      <div
        ref={elementRef}
        id={mountPointId}
      >
        {children}
      </div>
    </StyledBottomLayout>
  );
};

BottomLayoutComponent.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string
};

const mapStateToProps = ({ sequence }) => ({
  id: [sequence.currentTrainerIndex, sequence.started, sequence.completed].join()
});

export const BottomLayout = connect(mapStateToProps)(BottomLayoutComponent);

export const renderInBottomLayout = (WrappedComponent) => {
  const RenderInBottomLayout = (props) => {
    const [element, setElement] = useState(document.getElementById(mountPointId));

    useLayoutEffect(() => {
      const domNode = document.getElementById(mountPointId);
      if (!domNode) {
        // eslint-disable-next-line max-len
        console.warn(`DOM element with id ${mountPointId} not found. This is a requirement for a bottom layout to be properly rendered.`);
      }
      setElement(domNode);
    }, []);

    if (element === null) {
      return null;
    }

    return createPortal(
      <WrappedComponent {...props} />,
      element
    );
  };

  RenderInBottomLayout.displayName = `RenderInBottomLayout(${getDisplayName(WrappedComponent)})`;

  return RenderInBottomLayout;
};
