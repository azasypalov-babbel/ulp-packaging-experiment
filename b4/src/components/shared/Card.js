import { transparentize, rem } from 'polished';
import styled, { css } from 'styled-components';

const Card = styled.div`
  box-sizing: border-box;
  overflow: hidden;

  background: ${(props) => props.theme.cascada.white};

  border-radius: ${rem('4px')};

  transition: box-shadow 0.3s ease;

  box-shadow: 0 ${rem('2px')} ${rem('20px')} 0 ${(props) => transparentize(0.9, props.theme.cascada.spaceGrayW15)};

  outline: none;

  ${(props) => Boolean(props.onClick) && css`

    cursor: pointer;

    &:hover, &:focus {
      box-shadow: 0 ${rem('8px')} ${rem('40px')} 0 ${(props) => transparentize(0.8, props.theme.cascada.spaceGrayW15)};
    }

    &:active {
      box-shadow: 0 ${rem('2px')} ${rem('10px')} 0 ${(props) => transparentize(0.7, props.theme.cascada.spaceGrayW15)};
    }
  `}
`;

export default Card;
