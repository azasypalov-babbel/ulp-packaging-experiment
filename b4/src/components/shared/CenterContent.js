import styled from 'styled-components';

const CenterContent = styled.div`
   display: flex;
   align-items: center;

   & > * {
      margin: auto;
   }
`;

export default CenterContent;
