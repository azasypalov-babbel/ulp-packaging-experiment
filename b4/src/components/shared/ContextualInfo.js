import React, { useEffect, useRef } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { markupStringToHTML } from '@lessonnine/babbel-markup-helper.js';

import Text from './Text';
import { scrollIntoSafeArea } from './scrollIntoSafeArea';

const StyledWrapper = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-direction: column;

  width: 100%;
  height: auto;

  background-color: ${(props) => props.theme.color.silver};

  border-radius: 0.75rem;

  padding: 1rem;
`;

const ContextualInfo = ({
  markupString,
  active,
  ...props
}) => {
  const ref = useRef();

  useEffect(() => {
    if (active) {
      scrollIntoSafeArea(ref.current);
    }
  }, [ref, active]);

  return (
    <StyledWrapper ref={ref} data-item-type={props.dataItemType}>
      <Text
        data-text-type="contextual-info"
        data-selector="non-interactive"
        fontFamily="fontMilliardBook"
        color="spaceGray"
        textAlign="center"
        dangerouslySetInnerHTML={{ __html: markupStringToHTML(markupString) }}
      />
    </StyledWrapper>
  );
};

ContextualInfo.propTypes = {
  active: PropTypes.bool,
  markupString: PropTypes.string.isRequired,
  dataItemType: PropTypes.string
};

export default ContextualInfo;
