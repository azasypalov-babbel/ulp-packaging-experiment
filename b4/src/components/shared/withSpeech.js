import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getDisplayName } from '../getDisplayName';
import { requestMicPermissions } from '../../dux/permissions/actions';
import { withServices } from './withServices';
import rollbar from '../../services/rollbarService';
import { PERMISSIONS_STATUS } from '../../dux/permissions/status';

import * as Engines from '../../services/speechRecognition/engines';

export const SPEECHRECOGNITION_ERRORS = {
  NO_SPEECH: 'no-speech',
  ABORTED: 'aborted',
  AUDIO_CAPTURE: 'audio-capture',
  NETWORK: 'network',
  NOT_ALLOWED: 'not-allowed',
  SERVICE_NOT_ALLOWED: 'service-not-allowed',
  BAD_GRAMMAR: 'bad-grammar',
  LANGUAGE_NOT_SUPPORTED: 'language-not-supported'
};

export const CUSTOM_ERRORS = {
  UNKNOWN_ERROR: 'unknown-error',
  MIC_MUTED: 'mic-muted'
};

export const MEDIADEVICES_ERRORS = {
  NOT_ALLOWED: 'NotAllowedError'
};

const initialState = {
  confidenceScore: null,
  ended: false,
  error: '',
  listening: false,
  recording: false,
  solved: false,
  transcript: '',
  targetText: ''
};

const speechWordTimeout = 1000;

export const transcriptHasChanged = (prevSpeech, speech) =>
  speech.transcript !== prevSpeech.transcript &&
  speech.transcript !== initialState.transcript;

export const recognitionHasEnded = (prevSpeech, speech) =>
  prevSpeech.ended === false &&
  speech.ended === true;

export const recognitionHasError = (prevSpeech, speech) =>
  speech.error !== prevSpeech.error &&
  speech.error !== initialState.error;

/**
 * @deprecated in favour of `useSpeech`.
 */
export const withSpeech = (WrappedComponent) => {
  class WithSpeech extends React.Component {
    constructor(props) {
      super(props);

      this.state = initialState;

      this.handleStart = this.handleStart.bind(this);
      this.handleSpeechStart = this.handleSpeechStart.bind(this);
      this.handleEnd = this.handleEnd.bind(this);
      this.handleError = this.handleError.bind(this);
      this.handleResult = this.handleResult.bind(this);
      this.handleMutedChange = this.handleMutedChange.bind(this);

      this.start = this.start.bind(this);
      this.stop = this.stop.bind(this);
      this.reset = this.reset.bind(this);

      this.engineName = props.speechRecognitionService.getEngineName();
      this.advanceAutoSpeechStopTimeout = this.advanceAutoSpeechStopTimeout.bind(this);

      this.speechTimeout = null;
    }

    handleStart() {
      this.setState({
        confidenceScore: null,
        ended: false,
        error: '',
        recording: false,
        transcript: '',
        listening: true
      });
    }

    handleSpeechStart() {
      this.setState({ listening: false, recording: true });
    }

    handleEnd() {
      this.setState({ listening: false, recording: false, ended: true });
    }

    handleError({ error }) {
      this.setState({ error });
      const ignoreMessages = [
        SPEECHRECOGNITION_ERRORS.NO_SPEECH,
        SPEECHRECOGNITION_ERRORS.ABORTED
      ];
      if (error && !ignoreMessages.includes(error)) {
        rollbar.debug(`withSpeech SpeechRecognition Error: ${error}`);
      }
    }

    advanceAutoSpeechStopTimeout() {
      clearTimeout(this.speechTimeout);
      this.speechTimeout = setTimeout(this.stop, speechWordTimeout);
    }

    handleResult({
      transcript,
      isFinal,
      confidenceScore,
      solved
    }) {
      if (this.engineName === Engines.types.WEB_SPEECH && transcript) {
        this.advanceAutoSpeechStopTimeout();
      }

      this.setState({
        transcript,
        ended: isFinal,
        confidenceScore,
        solved
      });
    }

    handleMutedChange({ target }) {
      if (target.muted) {
        this.handleError({ error: CUSTOM_ERRORS.MIC_MUTED });
        this.stop();
      } else {
        this.handleError({ error: initialState.error });
      }
    }

    componentWillUnmount() {
      this.props.speechRecognitionService.cleanup();
      clearTimeout(this.speechTimeout);
    }

    componentDidMount() {
      this.props.requestMicPermissions();
    }

    start(options = {}) {
      this.setState({
        targetText: options.targetText
      });

      this.props.speechRecognitionService.start({
        onStart: this.handleStart,
        onSpeechStart: this.handleSpeechStart,
        onEnd: this.handleEnd,
        onError: this.handleError,
        onResult: this.handleResult,
        onMuteChange: this.handleMutedChange,
        locale: this.props.locale,
        learnLanguageAlpha3: this.props.learnLanguageAlpha3,
        ...options
      });
    }

    stop() {
      this.props.speechRecognitionService.stop();
    }

    reset() {
      this.setState({ ...initialState });
    }

    render() {
      const {
        confidenceScore,
        ended,
        error,
        listening,
        recording,
        solved,
        transcript
      } = this.state;

      const {
        permissionsGranted,
        ...restProps
      } = this.props;

      const speech = {
        permissionsGranted,
        confidenceScore,
        ended,
        error,
        listening,
        recording,
        solved,
        transcript,
        engineName: this.engineName,
        start: this.start,
        stop: this.stop,
        reset: this.reset
      };

      return (
        <WrappedComponent {...restProps} speech={speech} />
      );
    }
  }

  WithSpeech.displayName = `WithSpeech(${getDisplayName(WrappedComponent)})`;

  WithSpeech.propTypes = {
    locale: PropTypes.string.isRequired,
    learnLanguageAlpha3: PropTypes.string.isRequired,
    requestMicPermissions: PropTypes.func.isRequired,
    permissionsGranted: PropTypes.bool.isRequired,
    speechRecognitionService: PropTypes.any.isRequired
  };

  return WithSpeech;
};

const mapDispatchToProps = {
  requestMicPermissions
};

const mapStateToProps = ({ permissions, session }) => ({
  permissionsGranted: permissions.micPermission === PERMISSIONS_STATUS.granted,
  learnLanguageAlpha3: session.learnLanguageAlpha3,
  locale: session.locale
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withServices(['speechRecognitionService']),
  withSpeech
);

const speechEngineName = PropTypes.oneOf([
  Engines.types.WEB_SPEECH,
  Engines.types.NATIVE_SPEECH
]);

export const speechPropTypes = PropTypes.shape({
  permissionsGranted: PropTypes.bool.isRequired,
  ended: PropTypes.bool.isRequired,
  error: PropTypes.string.isRequired,
  listening: PropTypes.bool.isRequired,
  recording: PropTypes.bool.isRequired,
  reset: PropTypes.func.isRequired,
  start: PropTypes.func.isRequired,
  transcript: PropTypes.string.isRequired,
  solved: PropTypes.bool,
  engineName: speechEngineName
});
