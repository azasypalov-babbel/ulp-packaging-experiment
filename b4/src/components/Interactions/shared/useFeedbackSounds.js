import { useCallback, useContext, useEffect, useReducer, useRef } from 'react';
import { ServiceContext } from '@src/components/shared/withServices';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import { useSoundPlayer } from '@src/components/Trainer/shared/useSoundPlayer';
import { checks } from '@src/services/soundService';

const useFeedbackSounds = ({ skipOnActiveSound = true, resetDependencies = [] } = {}) => {
  const { settingsService, soundService } = useContext(ServiceContext);
  const correctSoundInstance = useSoundPlayer(correctFeedbackSound);
  const incorrectSoundInstance = useSoundPlayer(incorrectFeedbackSound);
  const [, forceUpdate] = useReducer((s) => !s, true);
  const skippedFeedbackRef = useRef(false);

  useEffect(() => {
    skippedFeedbackRef.current = false;
    correctSoundInstance.reset();
    incorrectSoundInstance.reset();
  }, [correctSoundInstance, incorrectSoundInstance, ...resetDependencies]);

  const hasPlayed = checks.hasPlayed(correctSoundInstance.getState())
    || checks.hasPlayed(incorrectSoundInstance.getState());

  return [
    useCallback((solved) => {
      const isPlayingAnySound = soundService.getPlayingInstances().length > 0;
      if ((skipOnActiveSound && isPlayingAnySound) || !settingsService.isFeedbackSoundEnabled) {
        skippedFeedbackRef.current = true;
        forceUpdate();
        return;
      }
      const instance = solved ? correctSoundInstance : incorrectSoundInstance;
      instance.play();
    }, [skipOnActiveSound, soundService, settingsService, correctSoundInstance, incorrectSoundInstance]),
    hasPlayed || skippedFeedbackRef.current
  ];
};

export default useFeedbackSounds;
