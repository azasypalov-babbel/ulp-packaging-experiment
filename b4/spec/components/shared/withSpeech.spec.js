import React from 'react';
import { shallow } from 'enzyme';
import { withSpeech } from '../../../src/components/shared/withSpeech';

const Component = () => <div />;
const WrappedComponent = withSpeech(Component);

const getUserMediaMock = jest.fn(() => Promise.resolve());

const speechRecognitionServiceMock = {
  start: jest.fn((options) => {
    Object.keys(options).forEach((prop) => {
      speechRecognitionServiceMock[prop] = options[prop];
    });
  }),
  stop: jest.fn(),
  cleanup: jest.fn(),
  getEngineName: jest.fn(() => 'WebAPISpeechRecognition')
};

const defaultProps = {
  speechRecognitionService: speechRecognitionServiceMock,
  requestMicPermissions: jest.fn(),
  permissionsGranted: true,
  locale: 'en',
  learnLanguageAlpha3: 'FRA'
};

describe('withSpeech HOC', () => {
  Object.defineProperty(global.navigator, 'mediaDevices', {
    value: {
      getUserMedia: getUserMediaMock
    }
  });

  beforeEach(() => {
    jest.clearAllMocks();
    global.AudioContext = jest.fn(() => ({ createMediaStreamSource: jest.fn() }));
  });

  test('passes speech prop to child component', () => {
    const wrapper = shallow(<WrappedComponent {...defaultProps} />);
    expect(wrapper.props().speech).toMatchSnapshot();
  });

  test('permissions are requested on mount', () => {
    shallow(<WrappedComponent {...defaultProps} />);
    expect(defaultProps.requestMicPermissions).toHaveBeenCalled();
  });

  describe('child calls start prop', () => {
    test('should start speech recognition service', () => {
      expect.assertions(1);

      const wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      expect(speechRecognitionServiceMock.start).toHaveBeenCalledWith(
        expect.objectContaining({
          locale: 'en',
          learnLanguageAlpha3: 'FRA'
        })
      );
    });

    test('should handle onStart event by making listening flag true', () => {
      const wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      expect.assertions(2);
      expect(wrapper.state('listening')).toEqual(false);
      speech.start();
      speechRecognitionServiceMock.onStart();
      expect(wrapper.state('listening')).toEqual(true);
    });
  });

  describe('no-speech error', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      speechRecognitionServiceMock.onStart();
      speechRecognitionServiceMock.onError({ error: 'no-speech' });
      speechRecognitionServiceMock.onEnd();
    });

    test('store error in state', () => {
      expect(wrapper.state('error')).toEqual('no-speech');
    });
  });

  describe('on speech end event', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      speechRecognitionServiceMock.onStart();
      speechRecognitionServiceMock.onEnd();
    });

    test('should set ended state and reset listening & recording flags', () => {
      expect(wrapper.state('ended')).toEqual(true);
      expect(wrapper.state('listening')).toEqual(false);
      expect(wrapper.state('recording')).toEqual(false);
    });
  });

  describe('speech with final result', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      speechRecognitionServiceMock.onStart();
    });

    test('transcript is updated in state', () => {
      speechRecognitionServiceMock.onResult({
        confidenceScore: 0.8,
        isFinal: false,
        transcript: 'die'
      });
      expect(wrapper.state('transcript')).toEqual('die');
      expect(wrapper.state('ended')).toEqual(false);

      speechRecognitionServiceMock.onResult({
        confidenceScore: 0.8,
        isFinal: false,
        transcript: 'die Stadt'
      });
      expect(wrapper.state('transcript')).toEqual('die Stadt');

      speechRecognitionServiceMock.onResult({
        confidenceScore: 0.8,
        isFinal: true,
        transcript: 'die Stadt'
      });

      expect(wrapper.state('ended')).toEqual(true);

      speechRecognitionServiceMock.onEnd();

      expect(wrapper.state('transcript')).toEqual('die Stadt');
      expect(wrapper.state('ended')).toEqual(true);
    });
  });

  describe('mute detection', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      speechRecognitionServiceMock.onStart();
    });
    describe('when mic is muted', () => {
      test('error MIC_MUTED is set into state', () => {
        speechRecognitionServiceMock.onMuteChange({
          target: { muted: true }
        });
        expect(wrapper.state('error')).toEqual('mic-muted');
      });
    });

    describe('when mic is unmuted', () => {
      test('should clear out error', () => {
        speechRecognitionServiceMock.onMuteChange({
          target: { muted: false }
        });
        expect(wrapper.state('error')).toEqual('');
      });
    });
  });

  describe('speech without final result', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      speechRecognitionServiceMock.onStart();
    });

    test('updates transcript in state', () => {
      speechRecognitionServiceMock.onResult({
        confidenceScore: 0.8,
        isFinal: false,
        transcript: 'die'
      });
      expect(wrapper.state('ended')).toEqual(false);

      speechRecognitionServiceMock.onEnd();

      expect(wrapper.state('transcript')).toEqual('die');
      expect(wrapper.state('ended')).toEqual(true);
    });
  });

  describe('Automatic speech timeout', () => {
    describe('when engine is WebAPISpeechRecognition', () => {
      let wrapper;
      beforeEach(() => {
        speechRecognitionServiceMock.getEngineName.mockImplementationOnce(() => 'WebAPISpeechRecognition');
        wrapper = shallow(<WrappedComponent {...defaultProps} />);
        const { speech } = wrapper.first().props();
        speech.start({
          targetText: 'Ich bin Sandra'
        });
      });

      test('Timeout is extended when result comes from speech recognition', () => {
        global.setTimeout = jest.fn();

        speechRecognitionServiceMock.onResult({
          confidenceScore: 0.8,
          isFinal: true,
          transcript: 'die'
        });

        const { speech } = wrapper.first().props();

        expect(global.setTimeout).toHaveBeenCalledWith(
          speech.stop,
          1000
        );
      });
    });

    describe('when engine is not WebAPISpeechRecognition', () => {
      let wrapper;
      beforeEach(() => {
        speechRecognitionServiceMock.getEngineName.mockImplementationOnce(() => 'OtherSpeechRecognition');
        wrapper = shallow(<WrappedComponent {...defaultProps} />);
        const { speech } = wrapper.first().props();
        speech.start({
          targetText: 'Ich bin Sandra'
        });
      });
      test('Timeout is never created', () => {
        global.setTimeout = jest.fn();

        speechRecognitionServiceMock.onResult({
          confidenceScore: 0.8,
          isFinal: true,
          transcript: 'die'
        });

        expect(global.setTimeout).not.toHaveBeenCalled();
      });
    });
  });

  describe('child calls stop prop', () => {
    test('it stops the speech recognition service', () => {
      expect.assertions(1);
      const wrapper = shallow(<WrappedComponent {...defaultProps} />);
      const { speech } = wrapper.first().props();
      speech.start();
      speech.stop();
      expect(speechRecognitionServiceMock.stop).toHaveBeenCalled();
    });
  });
});
