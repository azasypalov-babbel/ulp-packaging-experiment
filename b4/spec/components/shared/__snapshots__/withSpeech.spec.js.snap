// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSpeech HOC passes speech prop to child component 1`] = `
Object {
  "confidenceScore": null,
  "ended": false,
  "engineName": "WebAPISpeechRecognition",
  "error": "",
  "listening": false,
  "permissionsGranted": true,
  "recording": false,
  "reset": [Function],
  "solved": false,
  "start": [Function],
  "stop": [Function],
  "transcript": "",
}
`;
